core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -930,6 +962,69 @@ public void onFailure(Exception shardFailedError) {
         }
     }
 
+    /** a wrapper class to encapsulate a request when being sent to a specific allocation id **/
+    final class RequestWithAllocationID<R extends TransportRequest> extends TransportRequest {
```

Comment:
`ConcreteShardRequest` is much better than `RequestWithAllocationID` ? lets name classes about what they are not what properties they hold